File: pegasusX/apps/payload-terminal/__tests__/manifestBoard.test.ts
import { describe, expect, it } from 'vitest';
import {
  attachTruckStatus,
  BOARD_MANIFEST_STATES,
  groupBoardColumns,
  unassignedTrucks,
  type BoardTruck,
} from '../utils/manifestBoard';

describe('GS-U7 payload board', () => {
  it('groups trucks into four manifest-state columns', () => {
    const trucks: BoardTruck[] = [
      { id: 'd', label: 'D', license_plate: '', vehicle_class: 'TRUCK', truck_status: 'DRAFT' },
      { id: 'l', label: 'L', license_plate: '', vehicle_class: 'TRUCK', truck_status: 'LOADING' },
      { id: 's', label: 'S', license_plate: '', vehicle_class: 'TRUCK', truck_status: 'SEALED' },
      { id: 'x', label: 'X', license_plate: '', vehicle_class: 'TRUCK', truck_status: 'DISPATCHED' },
      { id: 'done', label: 'Z', license_plate: '', vehicle_class: 'TRUCK', truck_status: 'COMPLETED' },
      { id: 'none', label: 'N', license_plate: '', vehicle_class: 'TRUCK' },
    ];
    const cols = groupBoardColumns(trucks);
    expect(Object.keys(cols)).toEqual([...BOARD_MANIFEST_STATES]);
    expect(cols.DRAFT.map((t) => t.id)).toEqual(['d']);
    expect(cols.LOADING.map((t) => t.id)).toEqual(['l']);
    expect(cols.SEALED.map((t) => t.id)).toEqual(['s']);
    expect(cols.DISPATCHED.map((t) => t.id)).toEqual(['x']);
    expect(unassignedTrucks(trucks).map((t) => t.id)).toEqual(['done', 'none']);
  });

  it('attaches manifest state and does not invent DRAFT from COMPLETED', () => {
    const attached = attachTruckStatus(
      [
        { id: 'veh-1', label: 'A', license_plate: '', vehicle_class: 'TRUCK' },
        { id: 'veh-2', label: 'B', license_plate: '', vehicle_class: 'TRUCK' },
      ],
      [
        { vehicle_id: 'veh-1', state: 'SEALED', total_volume_vu: 8, max_volume_vu: 40, stop_count: 2 },
        { vehicle_id: 'veh-2', state: 'COMPLETED' },
      ],
    );
    expect(attached[0].truck_status).toBe('SEALED');
    expect(attached[0].used_volume_vu).toBe(8);
    expect(attached[1].truck_status ?? '').toBe('');
    const cols = groupBoardColumns(attached);
    expect(cols.SEALED).toHaveLength(1);
    expect(cols.DRAFT).toHaveLength(0);
  });

  it('empty board keeps four empty columns', () => {
    const cols = groupBoardColumns([]);
    expect(BOARD_MANIFEST_STATES.every((s) => cols[s].length === 0)).toBe(true);
  });
});

File: pegasusX/apps/payload-app-android/app/src/test/java/com/pegasus/payload/ui/home/ManifestBoardTest.kt
package com.pegasus.payload.ui.home

import com.pegasus.payload.data.model.Manifest
import com.pegasus.payload.data.model.Truck
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

class ManifestBoardTest {

    @Test
    fun boardHasFourManifestStateColumns() {
        val trucks = listOf(
            Truck(id = "d", truckStatus = "DRAFT"),
            Truck(id = "l", truckStatus = "LOADING"),
            Truck(id = "s", truckStatus = "SEALED"),
            Truck(id = "x", truckStatus = "DISPATCHED"),
            Truck(id = "done", truckStatus = "COMPLETED"),
            Truck(id = "none", truckStatus = ""),
        )
        val cols = ManifestBoard.group(trucks)
        assertEquals(listOf("DRAFT", "LOADING", "SEALED", "DISPATCHED"), cols.map { it.state })
        assertEquals(listOf("d"), cols[0].trucks.map { it.id })
        assertEquals(listOf("l"), cols[1].trucks.map { it.id })
        assertEquals(listOf("s"), cols[2].trucks.map { it.id })
        assertEquals(listOf("x"), cols[3].trucks.map { it.id })
        assertEquals(listOf("done", "none"), ManifestBoard.unassigned(trucks).map { it.id })
    }

    @Test
    fun attachUsesManifestStateNotInventedDraft() {
        val trucks = listOf(Truck(id = "veh-1"), Truck(id = "veh-2"))
        val manifests = listOf(
            Manifest(manifestId = "m1", vehicleId = "veh-1", state = "SEALED", totalVolumeVu = 8.0, maxVolumeVu = 40.0, stopCount = 2),
            Manifest(manifestId = "m2", vehicleId = "veh-2", state = "COMPLETED"),
        )
        val attached = ManifestBoard.attach(trucks, manifests)
        assertEquals("SEALED", attached[0].truckStatus)
        assertEquals(8L, attached[0].usedVolumeVu)
        assertEquals("", attached[1].truckStatus)
        val cols = ManifestBoard.group(attached)
        assertEquals(1, cols[2].trucks.size)
        assertTrue(cols[0].trucks.isEmpty())
    }

    @Test
    fun emptyBoardIsEmptyColumnsNotZeroTheatre() {
        val cols = ManifestBoard.group(emptyList())
        assertEquals(4, cols.size)
        assertTrue(cols.all { it.trucks.isEmpty() })
    }

    @Test
    fun pulseFailureDoesNotTreatAsEmptyTimeline() {
        val src = java.io.File("src/main/java/com/pegasus/payload/ui/home/HomeViewModel.kt").readText()
        assertTrue(src.contains("PulseHonesty.FAILED"))
        assertTrue(!src.contains("pulseEvents = emptyList()"))
        val screen = java.io.File("src/main/java/com/pegasus/payload/ui/home/HomeScreen.kt").readText()
        assertTrue(screen.contains("error = state.pulseError"))
    }
}
